src/crypto/public_key/elliptic/curves.js

```diff
@@ -152,75 +153,72 @@ function Curve(oid_or_name, params) {
   this.payloadSize = params.payloadSize;
-}
-
-Curve.prototype.keyFromPrivate = function (priv) { // Not for ed25519
-  return new KeyPair(this, { priv: priv });
-};
-
-Curve.prototype.keyFromPublic = function (pub) {
-  const keyPair = new KeyPair(this, { pub: pub });
-  if (
-    this.keyType === enums.publicKey.ecdsa &&
-    keyPair.keyPair.validate().result !== true
-  ) {
-    throw new Error('Invalid elliptic public key');
+  if (this.web && util.getWebCrypto()) {
+    this.type = 'web';
+  } else if (this.node && util.getNodeCrypto()) {
+    this.type = 'node';
+  } else if (this.name === 'curve25519') {
+    this.type = 'curve25519';
+  } else if (this.name === 'ed25519') {
+    this.type = 'ed25519';
   }
-  return keyPair;
-};
+}
 
 Curve.prototype.genKeyPair = async function () {
   let keyPair;
-  if (this.web && util.getWebCrypto()) {
-    // If browser doesn't support a curve, we'll catch it
-    try {
-      keyPair = await webGenKeyPair(this.name);
-    } catch (err) {
-      util.print_debug("Browser did not support signing: " + err.message);
+  switch (this.type) {
+    case 'web':
+      try {
+        return await webGenKeyPair(this.name);
+      } catch (err) {
+        util.print_debug_error("Browser did not support generating ec key " + err.message);
+        break;
+      }
+    case 'node':
+      return nodeGenKeyPair(this.name);
+    case 'curve25519': {
+      const privateKey = await random.getRandomBytes(32);
+      const one = new BN(1);
+      const mask = one.ushln(255 - 3).sub(one).ushln(3);
+      let secretKey = new BN(privateKey);
+      secretKey = secretKey.or(one.ushln(255 - 1));
+      secretKey = secretKey.and(mask);
+      secretKey = secretKey.toArrayLike(Uint8Array, 'le', 32);
+      keyPair = nacl.box.keyPair.fromSecretKey(secretKey);
+      const publicKey = util.concatUint8Array([new Uint8Array([0x40]), keyPair.publicKey]);
+      return { publicKey, privateKey };
     }
-  } else if (this.node && util.getNodeCrypto()) {
-    keyPair = await nodeGenKeyPair(this.name);
-  }
-
-  if (!keyPair || !keyPair.priv) {
-    // elliptic fallback
-    const r = await this.curve.genKeyPair({
-      entropy: util.Uint8Array_to_str(await random.getRandomBytes(32))
-    });
-    const compact = this.curve.curve.type === 'edwards' || this.curve.curve.type === 'mont';
-    if (this.keyType === enums.publicKey.eddsa) {
-      keyPair = { secret: r.getSecret() };
-    } else {
-      keyPair = { pub: r.getPublic('array', compact), priv: r.getPrivate().toArray() };
+    case 'ed25519': {
+      const privateKey = await random.getRandomBytes(32);
+      const keyPair = nacl.sign.keyPair.fromSeed(privateKey);
+      const publicKey = util.concatUint8Array([new Uint8Array([0x40]), keyPair.publicKey]);
+      return { publicKey, privateKey };
     }
   }
-  return new KeyPair(this, keyPair);
+  if (!util.getUseElliptic()) {
+    throw new Error('This curve is only supported in the full build of OpenPGP.js');
+  }
```

Comment:
I think you could move this check into the `getIndutnyCurve` function that you're calling below, and then also remove it in the other places you're checking this (since they will eventually all call `getIndutnyCurve`).

Reply:
If we move this check to the `getIndutnyCurve`, then sometimes in ecdh and ecdsa we would get rejected promise, and sometime we would get thrown error. Than we would need to catch those errors. It is considered anti-pattern.

```A function is hard to use when it sometimes returns a promise and sometimes throws an exception. When writing an async function, prefer to signal failure by returning a rejected promise```

https://stackoverflow.com/questions/33445415/javascript-promises-reject-vs-throw

Reply:
All of the functions are `async`, right? So it would always cause a rejected promise no matter where you put the `throw`.